### PR TITLE
Autocomplete elastic search

### PR DESCRIPTION
**Pull request by maggarwal13**

### CODE REVIEW

File: whois-api/src/main/java/net/ripe/db/whois/api/autocomplete/AutoCompleteSearch.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+import net.ripe.db.whois.common.rpsl.AttributeType;
+import net.ripe.db.whois.common.rpsl.ObjectType;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public interface AutoCompleteSearch {
```
Comment (eshryane) [inline whois-api/src/main/java/net/ripe/db/whois/api/autocomplete/AutoCompleteSearch.java:11]: Minor: autocomplete is one word so use ```AutocompleteSearch``` camel-casing. This is not easy to fix on OSX as the filename is case-insensitive

File: whois-api/src/main/java/net/ripe/db/whois/api/elasticsearch/ElasticFullTextIndex.java [hunk 2]
```diff
@@ -174,6 +168,25 @@ public Void extractData(final ResultSet rs) throws SQLException, DataAccessExcep
         LOGGER.info("Completed Rebuilding ES indexes");
     }
 
+    private boolean shouldRebuild() throws IOException {
+        if (elasticIndexService.getWhoisDocCount() == 0L) {
+            return true;
```
Comment (eshryane) [inline whois-api/src/main/java/net/ripe/db/whois/api/elasticsearch/ElasticFullTextIndex.java:173]: Add ```Logger.warn``` here to be consistent with the other conditions in that method

File: whois-api/src/test/java/net/ripe/db/whois/api/ElasticSearchHelper.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+                DeleteIndexRequest metadataRequest = new DeleteIndexRequest(metaDetaIndex);
+                esClient.indices().delete(metadataRequest, RequestOptions.DEFAULT);
+            } catch (Exception ignored) {
+            }
+        }
+    }
+
+    @NotNull
+    private static RestHighLevelClient getEsClient() {
+        return new RestHighLevelClient(RestClient.builder(new HttpHost(System.getProperty("elastic.host"), Integer.parseInt(System.getProperty("elastic.port")))));
```
Comment (eshryane) [inline whois-api/src/test/java/net/ripe/db/whois/api/ElasticSearchHelper.java:55]: Turn System.getProperty() calls into static fields at the top of the class so it's clearer (or make the class a Spring component and inject the values).

File: whois-api/src/test/java/net/ripe/db/whois/api/elasticsearch/AbstractElasticSearchIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+import org.junit.jupiter.api.BeforeEach;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.testcontainers.elasticsearch.ElasticsearchContainer;
+
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+
+public abstract class AbstractElasticSearchIntegrationTest extends AbstractIntegrationTest {
+
+    public static final String ENV_DISABLE_TEST_CONTAIENRS = "test.containers.disabled";
```
Comment (eshryane) [inline whois-api/src/test/java/net/ripe/db/whois/api/elasticsearch/AbstractElasticSearchIntegrationTest.java:23]: Typo: fix spelling of "CONTAINERS"

File: whois-api/src/test/java/net/ripe/db/whois/api/elasticsearch/AbstractElasticSearchIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+    @BeforeAll
+    public static void setUpElasticCluster() {
+        if (StringUtils.isBlank(System.getProperty(ENV_DISABLE_TEST_CONTAIENRS))) {
+            if (elasticsearchContainer == null) {
+                elasticsearchContainer = new ElasticsearchContainer("docker.elastic.co/elasticsearch/elasticsearch:7.15.0");
+                elasticsearchContainer.start();
+            }
+
+            System.setProperty("elastic.host", elasticsearchContainer.getHttpHostAddress().split(":")[0]);
```
Comment (eshryane) [inline whois-api/src/test/java/net/ripe/db/whois/api/elasticsearch/AbstractElasticSearchIntegrationTest.java:40]: Is this logic IPv6 specific? Can you separate the split logic into a separate method?

File: whois-api/src/main/java/net/ripe/db/whois/api/autocomplete/AutocompleteService.java [hunk 6]
```diff
@@ -29,15 +27,13 @@
 @Path("/autocomplete")
 public class AutocompleteService {
 
-    private static final Logger LOGGER = LoggerFactory.getLogger(AutocompleteService.class);
-
     private static final int MINIMUM_PREFIX_LENGTH = 2;
 
-    private final AutocompleteSearch autocompleteSearch;
+    private final AutocompleteSearch autoCompleteSearch;
```
Comment (eshryane) [inline whois-api/src/main/java/net/ripe/db/whois/api/autocomplete/AutocompleteService.java:32]: Fix camelcase, autocomplete is one word (I suggest changing to ```autocompleteSearch```)

File: whois-api/src/main/java/net/ripe/db/whois/api/autocomplete/ElasticAutocompleteSearch.java [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+@Component
+@Conditional(ElasticSearchCondition.class)
```
Comment (eshryane) [inline whois-api/src/main/java/net/ripe/db/whois/api/autocomplete/ElasticAutocompleteSearch.java:33]: Could you use ConditionalOnProperty instead? You can then drop the Conditional classes, may be simpler (less code)

```
@ConditionalOnProperty(prefix = "elasticsearch", name = "enabled", havingValue = "true")
```

File: whois-api/src/main/java/net/ripe/db/whois/api/autocomplete/LuceneAutocompleteSearch.java [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+@Component
+@Conditional(LuceneSearchCondition.class)
```
Comment (eshryane) [inline whois-api/src/main/java/net/ripe/db/whois/api/autocomplete/LuceneAutocompleteSearch.java:47]: Could you use ConditionalOnProperty instead? Then you can remove the Conditional classes. Define as the opposite to the ElasticSearch class.

```
@ConditionalOnProperty(prefix = "elasticsearch", name = "enabled", havingValue = "false", matchIfMissing="true")
```

File: whois-api/src/test/java/net/ripe/db/whois/api/ElasticSearchHelper.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+import org.elasticsearch.client.RestHighLevelClient;
+import org.elasticsearch.client.indices.CreateIndexRequest;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ElasticSearchHelper {
+    private static final Logger LOGGER = LoggerFactory.getLogger(DatabaseHelper.class);
+
+    public static final String ELASTIC_HOSTNAME = System.getProperty("elastic.host");
```
Comment (eshryane) [inline whois-api/src/test/java/net/ripe/db/whois/api/ElasticSearchHelper.java:17]: Can you make this class a Component and inject properties?

**Review by eshryane — COMMENTED** [review]
Minor comments

File: whois-api/src/main/java/net/ripe/db/whois/api/autocomplete/ElasticAutocompleteSearch.java (hunk 7 above)
Comment (maggarwal13) [reply whois-api/src/main/java/net/ripe/db/whois/api/autocomplete/ElasticAutocompleteSearch.java:33]: ConditionalOnProperty is springBoot annotation in spring we have only Conditional

File: whois-api/src/main/java/net/ripe/db/whois/api/autocomplete/LuceneAutocompleteSearch.java (hunk 8 above)
Comment (maggarwal13) [reply whois-api/src/main/java/net/ripe/db/whois/api/autocomplete/LuceneAutocompleteSearch.java:47]: Same as before: ConditionalOnProperty is springBoot annotation in spring we have only Conditional

File: whois-api/src/test/java/net/ripe/db/whois/api/ElasticSearchHelper.java (hunk 9 above)
Comment (maggarwal13) [reply whois-api/src/test/java/net/ripe/db/whois/api/ElasticSearchHelper.java:17]: Yes, initial idea was having it as a util class but it makes more sense to make a component now

File: whois-api/src/main/java/net/ripe/db/whois/api/autocomplete/LuceneSearchCondition.java [hunk 10]
```diff
@@ -0,0 +1,16 @@
+package net.ripe.db.whois.api.autocomplete;
+
+import org.springframework.context.annotation.Condition;
+import org.springframework.context.annotation.ConditionContext;
+import org.springframework.core.type.AnnotatedTypeMetadata;
+
+public class LuceneSearchCondition implements Condition {
```
Comment (eshryane) [inline whois-api/src/main/java/net/ripe/db/whois/api/autocomplete/LuceneSearchCondition.java:7]: Can you implement this class by extending ```ElasticSearchCondition``` and negating the super.matches() response? This eliminates the code duplication and guarantees that both conditions can never be true simultaneously

**Review by eshryane — APPROVED** [review]

**Pull request merged**